File: backend/src/main/java/com/zennyt/games/domain/config/MemoryQuestConfig.java
package com.zennyt.games.domain.config;

import java.util.List;

/**
 * Configuration du jeu « J'investigue » (GameType MEMORY_QUEST, mémoire de travail).
 *
 * <p>Java pur, sans Spring : règles issues du developer handoff « J'investigue ».
 * Le score est <b>calculé côté serveur</b> à partir des mesures brutes (le client
 * n'envoie jamais de points). Le mock mobile ({@code games_mock_repository.dart})
 * reproduit ce barème à l'identique.
 *
 * <p>Barème : chaque tâche (rappel même ordre, rappel inverse, restauration
 * d'objets, rappel après distraction) est notée <b>0–5</b> ; le composite est la
 * <b>moyenne des tâches jouées, normalisée /100</b> (indicatif, non diagnostique).
 */
public final class MemoryQuestConfig {

    private MemoryQuestConfig() {
    }

    /** Échelle de note par tâche (0–5). */
    public static final int TASK_MAX_SCORE = 5;

    /** Maximum du composite (/100). */
    public static final int COMPOSITE_MAX = 100;

    // ── Timings d'encodage (data-driven, handoff §6/§8) ──────────────────────
    public static final int DIGIT_VISIBLE_MS = 900;
    public static final int ISI_MS = 250;
    public static final int MANIPULATIONS_L1 = 2;
    public static final int DISTRACTION_SECONDS = 8; // 5–10 s

    // ── Système de niveaux (fiche « J'INVESTIGUE », Tableau 1) ────────────────

    /** {@code initial_sequence_length} — longueur de séquence au niveau 1. */
    public static final int INITIAL_SEQUENCE_LENGTH = 3;

    /** {@code sequence_increment} — +1 longueur par niveau. */
    public static final int SEQUENCE_INCREMENT = 1;

    /** {@code max_sequence_length} — longueur maximale testable. */
    public static final int MAX_SEQUENCE_LENGTH = 9;

    /** {@code total_levels} — 7 longueurs testables (3, 4, 5, 6, 7, 8, 9). */
    public static final int TOTAL_LEVELS = 7;

    /** {@code correct_tasks_for_level_up} — tâches réussies pour monter d'un niveau. */
    public static final int CORRECT_TASKS_FOR_LEVEL_UP = 3;

    /**
     * {@code reset_sequence_on_error} — sur erreur, la SÉQUENCE du prochain essai
     * est régénérée (nouvelle séquence), sans effacer le niveau déjà acquis. Le
     * crédit partiel (0–5) est calculé sur l'essai qui vient d'échouer,
     * indépendamment de ce reset.
     */
    public static final boolean RESET_SEQUENCE_ON_ERROR = true;

    /**
     * Nombre d'objets (Mission B) — un de plus par niveau à partir de
     * {@link #MIN_OBJECT_COUNT}, plafonné à {@link #MAX_OBJECT_COUNT}.
     *
     * <p>Le plancher est passé de 4 à <b>3</b> pour adoucir l'entrée dans le jeu
     * (demande client). Miroir de {@code memory_quest_config.dart}.
     */
    public static final int MIN_OBJECT_COUNT = 3;
    public static final int MAX_OBJECT_COUNT = 12;

    /**
     * {@code distraction_min_level} — la phase de distraction est <b>gatée</b> :
     * absente aux niveaux 1–2, active à partir du niveau 3.
     */
    public static final int DISTRACTION_MIN_LEVEL = 3;

    /** {@code hints_enabled} = false (mode évaluation : aucun indice). */
    public static final boolean HINTS_ENABLED = false;

    /** {@code partial_credit_enabled} = true (crédit partiel, cotation 0–5). */
    public static final boolean PARTIAL_CREDIT_ENABLED = true;

    /** {@code max_session_duration_min} — borne haute de la fiche. */
    public static final int MAX_SESSION_DURATION_MIN = 30;

    // ── Calibrage appareil → timeout par tâche (fiche Tableau 2) ─────────────
    // Memory Quest est le PREMIER module dont le score dépend du temps : le
    // socle DeviceCalibration / CalibrationService (déjà construit, non modifié)
    // est enfin exploité pour un score, pas seulement des indicateurs.

    /**
     * {@code max_task_time_ms} — délai max d'une tâche avant échec par dépassement.
     *
     * <p>// PROVISOIRE — à calibrer sur données pilotes (95ᵉ percentile observé).
     * La fiche indique qu'aucune valeur scientifique ne s'impose. Tracé dans
     * « Décisions à valider » (GAMES_MODULE.md).
     */
    public static final int MAX_TASK_TIME_MS = 6000;

    /**
     * {@code apply_calibration_to_task_timeout} — le délai est augmenté de
     * l'offset de calibrage avant de déclarer un échec, pour ne pas pénaliser un
     * appareil lent. C'est « l'ajustement serveur » du score final (fiche).
     */
    public static final boolean APPLY_CALIBRATION_TO_TASK_TIMEOUT = true;

    /**
     * Délai de timeout effectif, offset de calibrage compris : {@code max_task_time_ms
     * + calibration_offset_ms} (si l'ajustement est activé). L'offset provient de
     * {@code CalibrationService.offsetMs(deviceCalibration)} — socle réutilisé.
     *
     * @param calibrationOffsetMs offset technique de l'appareil (0 si aucun calibrage)
     */
    public static double adjustedTaskTimeoutMs(double calibrationOffsetMs) {
        double offset = APPLY_CALIBRATION_TO_TASK_TIMEOUT ? Math.max(0.0, calibrationOffsetMs) : 0.0;
        return MAX_TASK_TIME_MS + offset;
    }

    /**
     * true si une tâche est en échec par DÉPASSEMENT DE TEMPS (au-delà du timeout
     * ajusté du calibrage). Une tâche non échouée ici garde sa note de rappel ;
     * l'ajustement du calibrage ne fait que remonter le seuil pour un appareil lent.
     *
     * @param taskTimeMs          temps brut mesuré de la tâche
     * @param calibrationOffsetMs offset de calibrage de l'appareil
     */
    public static boolean isTaskTimedOut(long taskTimeMs, double calibrationOffsetMs) {
        return taskTimeMs > adjustedTaskTimeoutMs(calibrationOffsetMs);
    }

    // ── Validité de session (fiche Tableau 3) ────────────────────────────────

    /**
     * Seuil critique d'offset de calibrage : au-delà, l'appareil est jugé trop
     * peu fiable et la session est invalidée.
     *
     * <p>// PROVISOIRE — non chiffré par la fiche ; à calibrer sur données pilotes.
     * Tracé dans « Décisions à valider » (GAMES_MODULE.md).
     */
    public static final double CRITICAL_CALIBRATION_OFFSET_MS = 100.0;

    /**
     * Nombre max de tâches en échec par DÉPASSEMENT DE TEMPS toléré avant
     * d'invalider la session (au-delà, trop de timeouts vs erreurs de rappel).
     *
     * <p>// PROVISOIRE — non chiffré par la fiche ; à calibrer sur données pilotes.
     * Tracé dans « Décisions à valider » (GAMES_MODULE.md).
     */
    public static final int MAX_TIMEOUT_TASKS = 3;

    /**
     * {@code session_valid} — false si l'un des cas de la fiche (Tableau 3) :
     * <ol>
     *   <li>offset de calibrage &gt; {@link #CRITICAL_CALIBRATION_OFFSET_MS}
     *       (appareil trop peu fiable) ;</li>
     *   <li>la session n'atteint pas sa condition de fin (abandon) ;</li>
     *   <li>trop de tâches échouées par timeout (&gt; {@link #MAX_TIMEOUT_TASKS})
     *       plutôt que par erreur de rappel.</li>
     * </ol>
     *
     * @param calibrationOffsetMs offset de calibrage de l'appareil
     * @param sessionCompleted    true si la condition de fin a été atteinte
     * @param timeoutTaskCount    nb de tâches en échec par dépassement de temps
     */
    public static boolean isSessionValid(double calibrationOffsetMs,
                                         boolean sessionCompleted, int timeoutTaskCount) {
        if (calibrationOffsetMs > CRITICAL_CALIBRATION_OFFSET_MS) {
            return false; // (1) appareil jugé trop peu fiable
        }
        if (!sessionCompleted) {
            return false; // (2) abandon avant la fin
        }
        if (timeoutTaskCount > MAX_TIMEOUT_TASKS) {
            return false; // (3) trop de timeouts (dépassement) vs erreurs de rappel
        }
        return true;
    }

    /** Longueur de séquence attendue à un niveau (1-based) : {@code 3 + (level−1)}, plafonnée. */
    public static int sequenceLengthForLevel(int level) {
        int len = INITIAL_SEQUENCE_LENGTH + (Math.max(1, level) - 1) * SEQUENCE_INCREMENT;
        return Math.min(MAX_SEQUENCE_LENGTH, len);
    }

    /**
     * Nombre d'objets (Mission B) à un niveau (1-based) : <b>un objet de plus par
     * niveau</b> à partir de {@link #MIN_OBJECT_COUNT} — 3, 4, 5, 6… — plafonné à
     * {@link #MAX_OBJECT_COUNT}.
     *
     * <p>La progression était interpolée de bout en bout sur {@link #TOTAL_LEVELS},
     * ce qui donnait 4, 5, 7, 8, 9, 11, 12 : la charge sautait de deux objets à
     * certains paliers. Le pas constant reprend la mécanique du jeu de chiffres
     * (un chiffre de plus par niveau). Miroir de {@code memory_quest_config.dart}.
     */
    public static int objectCountForLevel(int level) {
        int count = MIN_OBJECT_COUNT + Math.max(1, level) - 1;
        return Math.min(count, MAX_OBJECT_COUNT);
    }

    // ── Difficulté des distractions visuelles (jeu des IMAGES) ───────────────
    //
    // Point de calibration UNIQUE de la montée en difficulté des tâches
    // parasites : durée, taille de grille, ressemblance, nombre de pièces.
    // Miroir de {@code memory_quest_config.dart}.

    /**
     * Niveau à partir duquel le jeu des <b>IMAGES</b> intercale une tâche
     * parasite — plus tôt que le jeu des chiffres
     * ({@link #DISTRACTION_MIN_LEVEL}), les deux courbes différant.
     */
    public static final int IMAGES_DISTRACTION_MIN_LEVEL = 2;

    public static boolean imagesDistractionActiveAtLevel(int level) {
        return level >= IMAGES_DISTRACTION_MIN_LEVEL;
    }

    /** Durée d'une tâche parasite au premier niveau où elle apparaît. */
    public static final int DISTRACTION_BASE_TIME_LIMIT_MS = 12_000;

    /** Temps retiré par niveau au-delà de {@link #DISTRACTION_MIN_LEVEL}. */
    public static final int DISTRACTION_TIME_STEP_MS = 500;

    /** Plancher : en deçà, la tâche cesse d'être faisable. */
    public static final int DISTRACTION_MIN_TIME_LIMIT_MS = 6_000;

    /**
     * Budget de temps d'une tâche parasite — <b>identique pour les deux
     * familles</b> (intrus / pièce manquante) et jamais illimité.
     */
    public static int distractionTimeLimitMs(int level) {
        int steps = Math.max(0, level - IMAGES_DISTRACTION_MIN_LEVEL);
        int ms = DISTRACTION_BASE_TIME_LIMIT_MS - steps * DISTRACTION_TIME_STEP_MS;
        return Math.max(ms, DISTRACTION_MIN_TIME_LIMIT_MS);
    }

    /** Cases de la grille « intrus » : 4, puis deux de plus par niveau. */
    public static final int DISTRACTION_MIN_GRID_CELLS = 4;
    public static final int DISTRACTION_MAX_GRID_CELLS = 12;

    public static int oddOneOutCellCount(int level) {
        int steps = Math.max(0, level - IMAGES_DISTRACTION_MIN_LEVEL);
        return Math.min(DISTRACTION_MIN_GRID_CELLS + steps * 2, DISTRACTION_MAX_GRID_CELLS);
    }

    /** Ressemblance de l'intrus dans [0,1] : plus haute = écart plus ténu. */
    public static final double DISTRACTION_BASE_SIMILARITY = 0.35;
    public static final double DISTRACTION_SIMILARITY_STEP = 0.10;
    public static final double DISTRACTION_MAX_SIMILARITY = 0.85;

    public static double oddOneOutSimilarity(int level) {
        int steps = Math.max(0, level - IMAGES_DISTRACTION_MIN_LEVEL);
        double s = DISTRACTION_BASE_SIMILARITY + steps * DISTRACTION_SIMILARITY_STEP;
        return Math.min(s, DISTRACTION_MAX_SIMILARITY);
    }

    /** Côté du puzzle : 2×2, puis un cran tous les deux niveaux, max 4×4. */
    public static final int PUZZLE_MIN_GRID_SIDE = 2;
    public static final int PUZZLE_MAX_GRID_SIDE = 4;

    public static int puzzleGridSide(int level) {
        int steps = Math.max(0, level - IMAGES_DISTRACTION_MIN_LEVEL);
        return Math.min(PUZZLE_MIN_GRID_SIDE + steps / 2, PUZZLE_MAX_GRID_SIDE);
    }

    /** Pièces proposées (une seule correcte) : 3, puis une de plus par niveau. */
    public static final int PUZZLE_MIN_OPTIONS = 3;
    public static final int PUZZLE_MAX_OPTIONS = 6;

    public static int puzzleOptionCount(int level) {
        int steps = Math.max(0, level - IMAGES_DISTRACTION_MIN_LEVEL);
        return Math.min(PUZZLE_MIN_OPTIONS + steps, PUZZLE_MAX_OPTIONS);
    }

    /**
     * Timeout d'une tâche PARASITE : c'est son propre budget qui fait foi, pas
     * {@link #MAX_TASK_TIME_MS}.
     */
    public static boolean isDistractionTimedOut(int taskTimeMs, int level, double offsetMs) {
        double limit = distractionTimeLimitMs(level)
            + (APPLY_CALIBRATION_TO_TASK_TIMEOUT ? Math.max(0, offsetMs) : 0);
        return taskTimeMs > limit;
    }

    /** true si la distraction est jouée à ce niveau (gatée à {@link #DISTRACTION_MIN_LEVEL}). */
    public static boolean distractionActiveAtLevel(int level) {
        return level >= DISTRACTION_MIN_LEVEL;
    }

    /** Note d'une tâche (0–5) à partir d'une précision [0,1] : {@code round(acc × 5)}. */
    public static int taskScore(double accuracy) {
        double clamped = Math.max(0.0, Math.min(1.0, accuracy));
        return (int) Math.round(clamped * TASK_MAX_SCORE);
    }

    // ── Interprétation du composite (bandes provisoires) ─────────────────────

    /** Bande d'interprétation : niveau attribué si {@code composite >= minInclusive}. */
    public record InterpretationBand(int minInclusive, String level) {
    }

    /**
     * Bandes d'interprétation du composite (/100).
     *
     * <p>// AJOUT NON VALIDÉ PAR LE PSYCHOLOGUE — bandes provisoires (mêmes seuils
     * indicatifs que les autres jeux). Le résultat est présenté comme
     * <b>indicatif, non diagnostique</b>.
     */
    public static final List<InterpretationBand> INTERPRETATION_BANDS = List.of(
        new InterpretationBand(90, "Excellent"),
        new InterpretationBand(75, "Bon"),
        new InterpretationBand(60, "Moyen"),
        new InterpretationBand(40, "Moyen faible"),
        new InterpretationBand(0, "Très faible")
    );

    /** Interprète un composite /100 selon les bandes provisoires ci-dessus. */
    public static String interpret(int composite) {
        return INTERPRETATION_BANDS.stream()
            .filter(band -> composite >= band.minInclusive())
            .map(InterpretationBand::level)
            .findFirst()
            .orElse("Très faible");
    }
}

File: mobile/lib/features/games/domain/config/memory_quest_config.dart
// Miroir Dart de `MemoryQuestConfig` (backend) — source UNIQUE côté mobile pour
// le système de niveaux, le timeout par tâche (calibrage) et la validité de
// session. L'écran et le mock lisent ces valeurs pour rester alignés backend ⇄ mock.

import 'dart:math' as math;

/// Configuration « J'investigue » (mémoire de travail) côté mobile.
class MemoryQuestConfig {
  MemoryQuestConfig._();

  static const int taskMaxScore = 5;
  static const int compositeMax = 100;

  // ── Système de niveaux (fiche Tableau 1) ──────────────────────────────────
  static const int initialSequenceLength = 3;
  static const int sequenceIncrement = 1;
  static const int maxSequenceLength = 9;
  static const int totalLevels = 7; // longueurs 3..9
  static const int correctTasksForLevelUp = 3;

  /// Un seul tour par niveau : après chaque tour on incrémente le niveau
  /// (séquence plus longue + plus d'objets), jusqu'au dernier niveau.
  static const int roundsPerLevel = 1;

  /// Échecs tolérés **dans un même niveau** avant la fin de la partie.
  ///
  /// Un tour raté ne fait pas monter de niveau : on rejoue le MÊME niveau avec
  /// une nouvelle séquence (cf. [resetSequenceOnError]). Au [maxFailuresPerLevel]
  /// ᵉ échec consécutif sur ce niveau, la partie s'arrête et l'écran de score
  /// s'affiche. Le compteur repart à zéro à chaque montée de niveau.
  ///
  /// Remplace un budget d'erreurs GLOBAL (3 sur toute la partie) qui ne
  /// correspondait pas au déroulé attendu : il additionnait des échecs de
  /// niveaux différents, et surtout le niveau montait même après un tour raté.
  static const int maxFailuresPerLevel = 2;
  static const bool resetSequenceOnError = true;
  /// Objets au premier niveau. Abaissé de 4 à **3** pour adoucir l'entrée dans
  /// le jeu, à la demande du client.
  static const int minObjectCount = 3;
  static const int maxObjectCount = 12;

  /// Niveau à partir duquel le jeu des **IMAGES** intercale une tâche parasite.
  ///
  /// Plus tôt que dans le jeu des chiffres ([distractionMinLevel]) : les deux
  /// jeux n'ont pas la même courbe, et le client veut voir les casse-tête
  /// visuels dès le second palier.
  static const int imagesDistractionMinLevel = 2;

  /// La tâche parasite visuelle est-elle due à ce niveau ?
  static bool imagesDistractionActiveAtLevel(int level) =>
      level >= imagesDistractionMinLevel;

  /// Niveau à partir duquel le jeu des **CHIFFRES** intercale sa question
  /// d'interférence (niveaux 1-2 sans, puis à CHAQUE niveau à partir de
  /// celui-ci).
  ///
  /// A brièvement valu 1 : le client signalait que le distracteur ne se voyait
  /// jamais, et le gating au niveau ≥ 3 en semblait la cause. La cause réelle
  /// était ailleurs — la distraction n'était accrochée qu'à la fin de la mission
  /// d'objets, donc absente du jeu des chiffres quel que soit le niveau (voir
  /// `_endLevelOrDistract` dans `investigate_screen.dart`). Ce point corrigé, le
  /// gating retrouve sa valeur de fiche, qui est aussi celle du backend
  /// (`MemoryQuestConfig.DISTRACTION_MIN_LEVEL = 3`) : les deux avaient divergé.
  static const int distractionMinLevel = 3;
  static const bool hintsEnabled = false;
  static const bool partialCreditEnabled = true;
  static const int maxSessionDurationMin = 30;

  /// Longueur de séquence à un niveau (1-based) : 3 + (level-1), plafonnée à 9.
  static int sequenceLengthForLevel(int level) {
    final len = initialSequenceLength + (level.clamp(1, 1 << 30) - 1) * sequenceIncrement;
    return len < maxSequenceLength ? len : maxSequenceLength;
  }

  /// Nombre d'objets (mémoire des images) à un niveau : **un objet de plus par
  /// niveau**, à partir de [minObjectCount] — 3, 4, 5, 6… plafonné à
  /// [maxObjectCount].
  ///
  /// La progression était auparavant interpolée de 4 à 12 sur les 7 niveaux, ce
  /// qui donnait 4, 5, 7, 8, 9, 11, 12 : la charge sautait de deux objets à
  /// certains paliers. Le pas constant reprend la mécanique du jeu de chiffres
  /// (un chiffre de plus par niveau), demandée par le client.
  static int objectCountForLevel(int level) {
    final count = minObjectCount + (level.clamp(1, 1 << 30) - 1);
    return count < maxObjectCount ? count : maxObjectCount;
  }

  /// Distraction active à partir de [distractionMinLevel].
  static bool distractionActiveAtLevel(int level) => level >= distractionMinLevel;

  // ── Temps d'observation des objets (Mission B) ─────────────────────────────
  //
  // MODÈLE : le temps accordé n'est PAS proportionnel au nombre d'objets. Il
  // suit une droite brisée au seuil de la mémoire de travail visuelle, estimée
  // à 4 ± 1 objets (Cowan, 2001) :
  //
  //   T(n) = base + perItem·n + rehearsal·max(0, n − capacity)
  //
  //   • `base`      — coût perceptif fixe : première saccade, lecture globale
  //                   de la grille, indépendant du nombre d'objets ;
  //   • `perItem`   — encodage d'une image. La littérature situe l'encodage
  //                   fiable d'une image entre 0,5 et 1 s (Potter, 1976 ;
  //                   Brady et al., PNAS 2008) ; le coût croît linéairement
  //                   avec le nombre d'éléments (Sternberg, Science 1966) ;
  //   • `rehearsal` — surcoût au-delà de la capacité : le joueur ne peut plus
  //                   tout tenir d'un coup et doit regrouper puis répéter
  //                   (Miller, 1956). C'est ce terme, absent d'un modèle
  //                   linéaire, qui rend les niveaux hauts jouables.
  //
  // Remplace un barème strictement proportionnel (1,25 s × n, plancher 5 s) qui
  // sur-dotait les petits lots — 5 s pour 3 objets, alors qu'ils tiennent en
  // mémoire de travail sans répétition — et sous-dotait les grands.

  /// Coût perceptif fixe, indépendant du nombre d'objets (ms).
  static const int objectObservationBaseMs = 2000;

  /// Encodage d'une image (ms), tant qu'on reste sous [workingMemoryCapacity].
  static const int objectObservationMsPerItem = 800;

  /// Surcoût de regroupement/répétition par objet AU-DELÀ de la capacité (ms).
  static const int objectObservationRehearsalMsPerItem = 500;

  /// Capacité de la mémoire de travail visuelle (Cowan, 2001) : le point où la
  /// pente change.
  static const int workingMemoryCapacity = 4;

  /// Plancher physiologique par objet (ms) : une fixation oculaire dure
  /// ~250-300 ms, donc en deçà le joueur n'a pas même le temps de poser les
  /// yeux sur chaque image. Ce plancher borne aussi l'ajustement individuel :
  /// sans lui, un joueur performant finirait par recevoir des niveaux
  /// littéralement infaisables.
  static const int objectObservationMinMsPerItem = 300;

  /// Plancher absolu du temps d'observation (ms).
  static const int objectObservationMinMs = 1500;

  /// Plafond du temps d'observation (ms). Au-delà, l'attention retombe : le
  /// joueur ne retient pas mieux, il s'ennuie.
  static const int objectObservationMaxMs = 30000;

  /// Temps d'observation des objets, en ms.
  ///
  /// [playerFactor] est le coefficient d'allure du joueur ([playerPaceNeutral]
  /// par défaut) — voir [nextPlayerFactor].
  ///
  /// Ex. à allure neutre : 3 objets → 4,4 s · 4 → 5,2 s · 6 → 7,8 s ·
  /// 9 → 11,7 s · 12 → 15,6 s.
  static int objectObservationMs(
    int objectCount, {
    double playerFactor = playerPaceNeutral,
  }) {
    final n = objectCount < 0 ? 0 : objectCount;
    final over = n - workingMemoryCapacity;
    final base = objectObservationBaseMs +
        objectObservationMsPerItem * n +
        objectObservationRehearsalMsPerItem * (over < 0 ? 0 : over);

    final scaled = (base * playerFactor).round();

    final fixationFloor = objectObservationMinMsPerItem * n;
    final floor = fixationFloor > objectObservationMinMs
        ? fixationFloor
        : objectObservationMinMs;

    if (scaled < floor) return floor;
    return scaled > objectObservationMaxMs ? objectObservationMaxMs : scaled;
  }

  // ── Temps de restitution (loi de Hick) ─────────────────────────────────────
  //
  // La restitution ne suit PAS la même loi que la mémorisation. Le joueur n'a
  // plus rien à encoder : il choisit, parmi les cartes affichées, laquelle
  // placer. Le temps de décision croît alors en logarithme du nombre d'options
  // (Hick, 1952 ; Hyman, 1953) :
  //
  //   T(n) = base + perDoubling · log2(n + 1)
  //
  // Réutiliser ici la droite de la mémorisation serait la faute classique : elle
  // rend les grands niveaux très généreux et les petits étouffants, alors que
  // c'est l'inverse qu'il faut. Ex. 3 objets → 5,9 s · 6 → 7,7 s · 9 → 8,8 s :
  // trois fois plus de cartes ne demandent pas trois fois plus de temps.

  /// Part fixe du temps de restitution (ms) : lecture de la consigne, premier
  /// balayage du plateau.
  static const int restoreBaseMs = 1500;

  /// Temps ajouté chaque fois que le nombre d'options double (ms).
  static const int restoreMsPerDoubling = 2200;

  /// Plancher absolu du temps de restitution (ms).
  static const int restoreMinMs = 4000;

  /// Plafond du temps de restitution (ms).
  static const int restoreMaxMs = 25000;

  /// Temps accordé pour reconstituer l'ordre, en ms.
  ///
  /// [playerFactor] est le même coefficient d'allure que la mémorisation : un
  /// joueur lent doit l'être sur les deux phases, sinon on lui rend d'un côté
  /// ce qu'on lui retire de l'autre.
  static int restoreTimeLimitMs(
    int objectCount, {
    double playerFactor = playerPaceNeutral,
  }) {
    final n = objectCount < 0 ? 0 : objectCount;
    final base =
        restoreBaseMs + restoreMsPerDoubling * (math.log(n + 1) / math.ln2);
    final scaled = (base * playerFactor).round();
    if (scaled < restoreMinMs) return restoreMinMs;
    return scaled > restoreMaxMs ? restoreMaxMs : scaled;
  }

  // ── Allure individuelle (escalier adaptatif) ───────────────────────────────
  //
  // Les constantes ci-dessus sont des moyennes de population ; un joueur donné
  // s'en écarte facilement de ±40 %. Le coefficient d'allure les recale, partie
  // après partie, par un escalier pondéré (Kaernbach, 1991) : le temps se réduit
  // d'un petit pas après un niveau réussi, et remonte d'un pas nettement plus
  // grand après un échec. Le rapport des deux pas fixe le point d'équilibre, ici
  // [targetSuccessRate] — le taux de réussite auquel l'apprentissage et
  // l'engagement sont maximaux (Wilson et al., Nature Communications, 2019).

  /// Coefficient neutre : le joueur reçoit exactement le barème de référence.
  static const double playerPaceNeutral = 1.0;

  /// Taux de réussite visé par l'escalier.
  static const double targetSuccessRate = 0.85;

  /// Pas de remontée après un échec, en log-temps. Règle la vitesse de
  /// convergence : plus haut = adaptation plus rapide mais plus instable.
  static const double playerPaceStepUp = 0.12;

  /// Bornes du coefficient. Elles garantissent qu'aucune série de réussites ne
  /// rende un niveau infaisable, ni qu'une série d'échecs ne le rende gratuit.
  static const double playerPaceMin = 0.6;
  static const double playerPaceMax = 1.8;

  /// Coefficient d'allure après un niveau, selon qu'il est réussi ou raté.
  ///
  /// Le pas de descente vaut `stepUp × (1 − p) / p` : c'est ce rapport qui fait
  /// converger l'escalier vers [targetSuccessRate] plutôt que vers 50 %.
  static double nextPlayerFactor(double current, {required bool success}) {
    final stepDown =
        playerPaceStepUp * (1 - targetSuccessRate) / targetSuccessRate;
    final next = current * math.exp(success ? -stepDown : playerPaceStepUp);
    if (next < playerPaceMin) return playerPaceMin;
    return next > playerPaceMax ? playerPaceMax : next;
  }

  // ── Difficulté des distractions visuelles (jeu des IMAGES) ────────────────
  //
  // Toute la montée en difficulté des tâches parasites est décrite ICI : durée,
  // taille de grille, ressemblance des éléments, nombre de pièces candidates.
  // Aucune de ces valeurs ne doit être réécrite ailleurs — c'est le point de
  // calibration unique.

  /// Durée d'une tâche parasite au premier niveau où elle apparaît.
  ///
  /// **Identique pour les deux types** (« intrus » et « pièce manquante ») :
  /// deux budgets de temps différents rendraient les niveaux incomparables.
  static const int distractionBaseTimeLimitMs = 12000;

  /// Temps retiré à chaque niveau au-delà de [distractionMinLevel].
  static const int distractionTimeStepMs = 500;

  /// Plancher : en deçà, la tâche cesse d'être faisable et ne mesure plus rien.
  static const int distractionMinTimeLimitMs = 6000;

  /// Budget de temps d'une tâche parasite à un niveau donné — **jamais illimité**.
  static int distractionTimeLimitMs(int level) {
    final steps = (level - imagesDistractionMinLevel).clamp(0, 1 << 30);
    final ms = distractionBaseTimeLimitMs - steps * distractionTimeStepMs;
    return ms < distractionMinTimeLimitMs ? distractionMinTimeLimitMs : ms;
  }

  /// Nombre de cases de la grille « trouver l'intrus » : 4 au premier niveau,
  /// deux de plus par niveau, plafonné à [distractionMaxGridCells].
  static const int distractionMinGridCells = 4;
  static const int distractionMaxGridCells = 12;

  static int oddOneOutCellCount(int level) {
    final steps = (level - imagesDistractionMinLevel).clamp(0, 1 << 30);
    final cells = distractionMinGridCells + steps * 2;
    return cells < distractionMaxGridCells ? cells : distractionMaxGridCells;
  }

  /// Ressemblance entre l'intrus et les autres cases, dans [0, 1] : plus elle
  /// est haute, plus l'écart visuel est ténu — donc plus la recherche est
  /// longue. 0,35 au premier niveau, +0,1 par niveau, plafonnée.
  static const double distractionBaseSimilarity = 0.35;
  static const double distractionSimilarityStep = 0.10;
  static const double distractionMaxSimilarity = 0.85;

  static double oddOneOutSimilarity(int level) {
    final steps = (level - imagesDistractionMinLevel).clamp(0, 1 << 30);
    final s = distractionBaseSimilarity + steps * distractionSimilarityStep;
    return s > distractionMaxSimilarity ? distractionMaxSimilarity : s;
  }

  /// Côté de la grille du puzzle : 2×2, puis 3×3, puis 4×4 — un cran tous les
  /// deux niveaux.
  static const int puzzleMinGridSide = 2;
  static const int puzzleMaxGridSide = 4;

  static int puzzleGridSide(int level) {
    final steps = (level - imagesDistractionMinLevel).clamp(0, 1 << 30);
    final side = puzzleMinGridSide + steps ~/ 2;
    return side > puzzleMaxGridSide ? puzzleMaxGridSide : side;
  }

  /// Pièces proposées pour combler le trou (une seule est correcte) : 3 au
  /// premier niveau, une de plus par niveau.
  static const int puzzleMinOptions = 3;
  static const int puzzleMaxOptions = 6;

  static int puzzleOptionCount(int level) {
    final steps = (level - imagesDistractionMinLevel).clamp(0, 1 << 30);
    final n = puzzleMinOptions + steps;
    return n > puzzleMaxOptions ? puzzleMaxOptions : n;
  }

  // ── Calibrage appareil → timeout par tâche (fiche Tableau 2) ──────────────
  /// PROVISOIRE — à calibrer sur données pilotes (95ᵉ percentile). Miroir backend.
  static const int maxTaskTimeMs = 6000;
  static const bool applyCalibrationToTaskTimeout = true;

  /// Timeout effectif, offset de calibrage compris.
  static double adjustedTaskTimeoutMs(double offsetMs) =>
      maxTaskTimeMs + (applyCalibrationToTaskTimeout ? (offsetMs < 0 ? 0 : offsetMs) : 0);

  static bool isTaskTimedOut(int taskTimeMs, double offsetMs) =>
      taskTimeMs > adjustedTaskTimeoutMs(offsetMs);

  /// Timeout d'une tâche PARASITE : c'est son propre budget qui fait foi, pas
  /// [maxTaskTimeMs].
  ///
  /// Sans cette distinction, une tâche parasite résolue en 8 s — dans les temps
  /// vis-à-vis de son chronomètre de 12 s — serait quand même voidée par le
  /// seuil générique de 6 s. Le joueur serait puni deux fois pour un essai
  /// réussi.
  static bool isDistractionTimedOut(int taskTimeMs, int level, double offsetMs) {
    final limit = distractionTimeLimitMs(level) +
        (applyCalibrationToTaskTimeout ? (offsetMs < 0 ? 0 : offsetMs) : 0);
    return taskTimeMs > limit;
  }

  // ── Validité de session (fiche Tableau 3) — seuils PROVISOIRES ─────────────
  static const double criticalCalibrationOffsetMs = 100;
  static const int maxTimeoutTasks = 3;

  static bool isSessionValid(double offsetMs, bool sessionCompleted, int timeoutTaskCount) {
    if (offsetMs > criticalCalibrationOffsetMs) return false;
    if (!sessionCompleted) return false;
    if (timeoutTaskCount > maxTimeoutTasks) return false;
    return true;
  }

  /// Note d'une tâche (0–5) à partir d'une précision [0,1].
  static int taskScore(double accuracy) =>
      (accuracy.clamp(0.0, 1.0) * taskMaxScore).round();
}
